File: scripts/qemu_verify/udf.py
"""Minimal UDF reader used by verify-qemu-image."""

from __future__ import annotations

from .common import FileExtent, FileRecord, Partition, require, u16, u32, u64


class UdfVolume:
    fs_type = "udf"

    def __init__(self, image, partition: Partition):
        self.image = image
        self.partition = partition
        self.block_size = image.sector_size
        self.partition_start = 0
        self.root_icb = 0
        self.mount()

    def read_block(self, lba: int) -> bytes:
        return self.image.read_blocks(self.partition.start_lba + lba)

    def read_partition_block(self, block_num: int) -> bytes:
        return self.read_block(self.partition_start + block_num)

    def mount(self) -> None:
        anchor = None
        for lba in (256, 512):
            if lba >= self.partition.block_count:
                continue
            block = self.read_block(lba)
            if u16(block, 0) == 0x0002 and u32(block, 12) == lba:
                anchor = block
                break
        require(anchor is not None, f"{self.partition.name}: missing UDF anchor")

        vds_start = u32(anchor, 20)
        vds_blocks = max(1, (u32(anchor, 16) + self.block_size - 1) // self.block_size)
        root_hint = 0
        partition_number = None
        for lba in range(vds_start, vds_start + vds_blocks):
            block = self.read_block(lba)
            ident = u16(block, 0)
            if ident == 0x0005:
                partition_number = u16(block, 22)
                self.partition_start = u32(block, 188)
            elif ident == 0x0006:
                require(u32(block, 212) == self.block_size, f"{self.partition.name}: UDF block size mismatch")
                root_hint = u32(block, 252)
                require(block[440] == 1 and block[441] >= 6, f"{self.partition.name}: unsupported UDF partition map")
                require(partition_number is None or u16(block, 444) == partition_number, f"{self.partition.name}: UDF map mismatch")
            elif ident == 0x0008:
                break

        require(self.partition_start > 0, f"{self.partition.name}: missing UDF partition descriptor")
        fsd = self.read_partition_block(root_hint)
        require(u16(fsd, 0) == 0x0100, f"{self.partition.name}: missing UDF file set descriptor")
        self.root_icb = u32(fsd, 404)

    def read_icb(self, block_num: int) -> tuple[bytes, int]:
        block = self.read_partition_block(block_num)
        ident = u16(block, 0)
        require(ident in (0x0105, 0x010A), f"{self.partition.name}: invalid UDF file entry")
        return block, ident

    def node_info(self, block_num: int) -> tuple[bool, int, int, int, int]:
        entry, ident = self.read_icb(block_num)
        is_dir = entry[27] == 0x04
        is_file = entry[27] == 0x05
        require(is_dir or is_file, f"{self.partition.name}: unsupported UDF node type")
        if ident == 0x0105:
            size = u64(entry, 56)
            alloc_offset = 176 + u32(entry, 168)
            alloc_len = u32(entry, 172)
        else:
            size = u64(entry, 64)
            alloc_offset = 216 + u32(entry, 208)
            alloc_len = u32(entry, 212)
        return is_dir, size, u16(entry, 34), alloc_offset, alloc_len

    def decode_name(self, raw: bytes) -> str:
        require(raw, f"{self.partition.name}: empty UDF name")
        if raw[0] == 8:
            return raw[1:].decode("utf-8")
        if raw[0] == 16:
            return raw[1:].decode("utf-16-be")
        raise AssertionError(f"{self.partition.name}: unsupported UDF name compression {raw[0]}")

    def file_extents(self, record: FileRecord) -> list[FileExtent]:
        is_dir, size, flags, alloc_offset, alloc_len = self.node_info(record.first_cluster)
        require(not is_dir, f"{record.name}: expected UDF file")
        require(flags & 0x0007 == 0, f"{record.name}: unsupported UDF allocation descriptor")
        if size == 0:
            return []
        entry, _ = self.read_icb(record.first_cluster)
        extents: list[FileExtent] = []
        virtual_block = 0
        for offset in range(alloc_offset, alloc_offset + alloc_len, 8):
            raw_len = u32(entry, offset)
            length = raw_len & 0x3FFFFFFF
            extent_type = raw_len & 0xC0000000
            block_num = u32(entry, offset + 4)
            if length == 0:
                continue
            blocks = (length + self.block_size - 1) // self.block_size
            if extent_type == 0:
                extents.append(FileExtent(virtual_block, self.partition.start_lba + self.partition_start + block_num, blocks))
            virtual_block += blocks
        return extents

    def read_directory(self, block_num: int) -> list[FileRecord]:
        is_dir, size, flags, alloc_offset, alloc_len = self.node_info(block_num)
        require(is_dir, f"{self.partition.name}: UDF node is not a directory")
        require(flags & 0x0007 == 0, f"{self.partition.name}: unsupported UDF directory descriptor")
        entry, _ = self.read_icb(block_num)
        require(alloc_len >= 8, f"{self.partition.name}: UDF directory has no data")
        dir_block = u32(entry, alloc_offset + 4)
        data = self.read_partition_block(dir_block)

        records: list[FileRecord] = []
        offset = 0
        while offset + 38 <= size:
            require(u16(data, offset) == 0x0101, f"{self.partition.name}: invalid UDF FID")
            characteristics = data[offset + 18]
            name_len = data[offset + 19]
            icb_block = u32(data, offset + 24)
            imp_use_len = u16(data, offset + 36)
            name_start = offset + 38 + imp_use_len
            name_end = name_start + name_len
            raw_name = data[name_start:name_end]
            if characteristics & (0x04 | 0x08) == 0:
                child_is_dir, child_size, _, _, _ = self.node_info(icb_block)
                records.append(FileRecord(self.decode_name(raw_name), child_is_dir, child_size, icb_block, True))
            offset = (name_end + 3) & ~3
        return records

    def lookup(self, path: str) -> FileRecord:
        parts = [part for part in path.replace("\\", "/").split("/") if part]
        record = FileRecord("/", True, 0, self.root_icb, True)
        for index, part in enumerate(parts):
            entries = self.read_directory(record.first_cluster)
            for entry in entries:
                if entry.name.lower() == part.lower():
                    record = entry
                    break
            else:
                raise AssertionError(f"{self.partition.name}: missing UDF path /{'/'.join(parts[:index + 1])}")
        return record

File: crates/nextboot-fs/src/udf.rs
//! Minimal read-only UDF filesystem support.
//!
//! This implements the subset needed by UEFI boot paths on hybrid Windows
//! installation media: type-1 partition maps, FE/EFE file entries, short/long
//! allocation descriptors, in-ICB data, and File Identifier Descriptor
//! directories.

use crate::{
    alloc_buffer, FileAttributes, FileExtent, FileInfo, FileSystem, FileSystemType, FsError,
    SharedBlockIo,
};
use alloc::string::String;
use alloc::vec::Vec;

mod mount;
mod nodes;
mod storage;

const TAG_IDENT_AVDP: u16 = 0x0002;
const TAG_IDENT_PD: u16 = 0x0005;
const TAG_IDENT_LVD: u16 = 0x0006;
const TAG_IDENT_TD: u16 = 0x0008;
const TAG_IDENT_FSD: u16 = 0x0100;
const TAG_IDENT_FID: u16 = 0x0101;
const TAG_IDENT_FE: u16 = 0x0105;
const TAG_IDENT_EFE: u16 = 0x010a;

const ICB_FILE_TYPE_DIRECTORY: u8 = 0x04;
const ICB_FILE_TYPE_REGULAR: u8 = 0x05;

const ICB_AD_SHORT: u16 = 0x0000;
const ICB_AD_LONG: u16 = 0x0001;
const ICB_AD_EXTENDED: u16 = 0x0002;
const ICB_AD_IN_ICB: u16 = 0x0003;
const ICB_AD_MASK: u16 = 0x0007;

const EXTENT_TYPE_MASK: u32 = 0xc000_0000;
const EXTENT_LENGTH_MASK: u32 = 0x3fff_ffff;

const FID_CHAR_HIDDEN: u8 = 0x01;
const FID_CHAR_DIRECTORY: u8 = 0x02;
const FID_CHAR_DELETED: u8 = 0x04;
const FID_CHAR_PARENT: u8 = 0x08;

const AVDP_CANDIDATES: &[u64] = &[256, 512];
const TAG_IDENT_OFFSET: usize = 0;
const TAG_LOCATION_OFFSET: usize = 12;
const AVDP_MAIN_VDS_LENGTH_OFFSET: usize = 16;
const AVDP_MAIN_VDS_START_OFFSET: usize = 20;
const PD_PARTITION_NUMBER_OFFSET: usize = 22;
const PD_PARTITION_START_OFFSET: usize = 188;
const PD_PARTITION_LENGTH_OFFSET: usize = 192;
const LVD_BLOCK_SIZE_OFFSET: usize = 212;
const LVD_ROOT_FILESET_OFFSET: usize = 248;
const LVD_MAP_TABLE_LENGTH_OFFSET: usize = 264;
const LVD_NUM_PARTITION_MAPS_OFFSET: usize = 268;
const LVD_PARTITION_MAPS_OFFSET: usize = 440;
const FSD_ROOT_ICB_OFFSET: usize = 400;
const FILE_ENTRY_ICB_FILE_TYPE_OFFSET: usize = 27;
const FILE_ENTRY_ICB_FLAGS_OFFSET: usize = 34;
const FILE_ENTRY_FILE_SIZE_OFFSET: usize = 56;
const FE_BLOCKS_RECORDED_OFFSET: usize = 64;
const FE_EXT_ATTR_LENGTH_OFFSET: usize = 168;
const FE_ALLOC_DESCS_LENGTH_OFFSET: usize = 172;
const FE_ALLOC_DESCS_OFFSET: usize = 176;
const EFE_OBJECT_SIZE_OFFSET: usize = 64;
const EFE_BLOCKS_RECORDED_OFFSET: usize = 72;
const EFE_EXT_ATTR_LENGTH_OFFSET: usize = 208;
const EFE_ALLOC_DESCS_LENGTH_OFFSET: usize = 212;
const EFE_ALLOC_DESCS_OFFSET: usize = 216;
const FID_HEADER_SIZE: usize = 38;
const FID_CHARACTERISTICS_OFFSET: usize = 18;
const FID_NAME_LENGTH_OFFSET: usize = 19;
const FID_ICB_OFFSET: usize = 20;
const FID_IMP_USE_LENGTH_OFFSET: usize = 36;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
struct ExtentAd {
    length: u32,
    start: u32,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
struct LogicalBlockAddress {
    block_num: u32,
    part_ref: u16,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
struct LongAd {
    length: u32,
    block: LogicalBlockAddress,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
struct Partition {
    number: u16,
    start: u32,
    length: u32,
    descriptor_lba: u64,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
struct PartitionMap {
    partition_index: usize,
}

#[derive(Debug, Clone)]
struct UdfNode {
    entry_lba: u64,
    tag_ident: u16,
    part_ref: u16,
    file_type: u8,
    flags: u16,
    file_size: u64,
    alloc_desc_offset: usize,
    alloc_desc_len: usize,
    entry: Vec<u8>,
}

impl UdfNode {
    fn is_dir(&self) -> bool {
        self.file_type == ICB_FILE_TYPE_DIRECTORY
    }

    fn is_file(&self) -> bool {
        self.file_type == ICB_FILE_TYPE_REGULAR
    }
}

/// A block-level patch that redirects a UDF file entry to replacement data.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct UdfFileReplacementPatch {
    pub file_entry_offset: u64,
    pub file_entry_data: Vec<u8>,
    pub partition_descriptor: Option<UdfPartitionDescriptorPatch>,
}

/// A patched UDF partition descriptor block, needed when appended replacement
/// data extends past the original partition extent.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct UdfPartitionDescriptorPatch {
    pub descriptor_offset: u64,
    pub descriptor_data: Vec<u8>,
}

/// Read-only UDF filesystem.
pub struct Udf {
    block_io: SharedBlockIo,
    block_size: u32,
    logical_block_size: u32,
    partitions: Vec<Partition>,
    partition_maps: Vec<PartitionMap>,
    root_icb: LongAd,
}

impl FileSystem for Udf {
    const FS_TYPE: FileSystemType = FileSystemType::Udf;

    fn init(block_io: SharedBlockIo) -> Result<Self, FsError> {
        let block_size = block_io.block_size();
        if block_size == 0 {
            return Err(FsError::InvalidArgument);
        }

        let mut fs = Self {
            block_io,
            block_size,
            logical_block_size: block_size,
            partitions: Vec::new(),
            partition_maps: Vec::new(),
            root_icb: LongAd {
                length: 0,
                block: LogicalBlockAddress {
                    block_num: 0,
                    part_ref: 0,
                },
            },
        };
        fs.mount()?;
        Ok(fs)
    }

    fn read_dir(&self, path: &str) -> Result<Vec<FileInfo>, FsError> {
        let node = self.find_node(path)?;
        if !node.is_dir() {
            return Err(FsError::NotDirectory);
        }

        self.read_dir_node(&node)
    }

    fn read_file(&self, path: &str, offset: u64, buf: &mut [u8]) -> Result<usize, FsError> {
        let node = self.find_node(path)?;
        if !node.is_file() {
            return Err(FsError::NotFile);
        }

        self.read_node_data(&node, offset, buf)
    }

    fn stat(&self, path: &str) -> Result<FileInfo, FsError> {
        if path == "/" || path.is_empty() {
            return Ok(FileInfo::new(String::from("/"), 0, true, 0));
        }

        let (dir, name) = crate::split_path(path);
        let entries = self.read_dir(&dir)?;
        entries
            .into_iter()
            .find(|entry| entry.name.eq_ignore_ascii_case(&name))
            .ok_or(FsError::FileNotFound)
    }

    fn block_size(&self) -> u32 {
        self.logical_block_size
    }

    fn file_extents(&self, path: &str) -> Result<Vec<FileExtent>, FsError> {
        let node = self.find_node(path)?;
        if !node.is_file() {
            return Err(FsError::NotFile);
        }

        self.node_extents(&node)
    }
}

impl Udf {
    /// Open a UDF filesystem from a shared block device.
    pub fn open(block_io: SharedBlockIo) -> Result<Self, FsError> {
        <Self as FileSystem>::init(block_io)
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
struct UdfDirEntry {
    name: String,
    icb: LongAd,
    is_dir: bool,
    hidden: bool,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
struct NodeExtent {
    length: u32,
    physical_lba: u64,
    extent_type: u32,
}

fn read_long_ad(data: &[u8], offset: usize) -> Result<LongAd, FsError> {
    Ok(LongAd {
        length: read_u32(data, offset)?,
        block: LogicalBlockAddress {
            block_num: read_u32(data, offset + 4)?,
            part_ref: read_u16(data, offset + 8)?,
        },
    })
}

fn decode_osta_name(raw: &[u8]) -> Result<String, FsError> {
    let Some((&compression, data)) = raw.split_first() else {
        return Err(FsError::Corrupted);
    };

    let mut out = String::new();
    match compression {
        8 => {
            out.try_reserve(data.len())
                .map_err(|_| FsError::OutOfMemory)?;
            for &byte in data {
                out.push(byte as char);
            }
        }
        16 => {
            if data.len() % 2 != 0 {
                return Err(FsError::Corrupted);
            }
            out.try_reserve(data.len() / 2)
                .map_err(|_| FsError::OutOfMemory)?;
            for unit in data.chunks_exact(2) {
                let ch = u16::from_be_bytes([unit[0], unit[1]]);
                out.push(char::from_u32(u32::from(ch)).unwrap_or('\u{fffd}'));
            }
        }
        _ => return Err(FsError::UnsupportedFs),
    }

    Ok(out)
}

fn read_u16(data: &[u8], offset: usize) -> Result<u16, FsError> {
    let bytes = data.get(offset..offset + 2).ok_or(FsError::Corrupted)?;
    Ok(u16::from_le_bytes([bytes[0], bytes[1]]))
}

fn read_u32(data: &[u8], offset: usize) -> Result<u32, FsError> {
    let bytes = data.get(offset..offset + 4).ok_or(FsError::Corrupted)?;
    Ok(u32::from_le_bytes([bytes[0], bytes[1], bytes[2], bytes[3]]))
}

fn read_u64(data: &[u8], offset: usize) -> Result<u64, FsError> {
    let bytes = data.get(offset..offset + 8).ok_or(FsError::Corrupted)?;
    Ok(u64::from_le_bytes([
        bytes[0], bytes[1], bytes[2], bytes[3], bytes[4], bytes[5], bytes[6], bytes[7],
    ]))
}

fn write_u16(data: &mut [u8], offset: usize, value: u16) -> Result<(), FsError> {
    let bytes = data.get_mut(offset..offset + 2).ok_or(FsError::Corrupted)?;
    bytes.copy_from_slice(&value.to_le_bytes());
    Ok(())
}

fn write_u32(data: &mut [u8], offset: usize, value: u32) -> Result<(), FsError> {
    let bytes = data.get_mut(offset..offset + 4).ok_or(FsError::Corrupted)?;
    bytes.copy_from_slice(&value.to_le_bytes());
    Ok(())
}

fn write_u64(data: &mut [u8], offset: usize, value: u64) -> Result<(), FsError> {
    let bytes = data.get_mut(offset..offset + 8).ok_or(FsError::Corrupted)?;
    bytes.copy_from_slice(&value.to_le_bytes());
    Ok(())
}

fn refresh_descriptor_tag(block: &mut [u8]) -> Result<(), FsError> {
    if block.len() < 16 {
        return Err(FsError::Corrupted);
    }

    let crc_len = read_u16(block, 10)? as usize;
    if crc_len > 0 {
        let crc_end = 16usize.checked_add(crc_len).ok_or(FsError::Corrupted)?;
        if crc_end > block.len() {
            return Err(FsError::Corrupted);
        }
        let crc = udf_crc16(&block[16..crc_end]);
        write_u16(block, 8, crc)?;
    }

    block[4] = 0;
    let checksum = block[..16]
        .iter()
        .enumerate()
        .filter(|(index, _)| *index != 4)
        .fold(0u8, |sum, (_, byte)| sum.wrapping_add(*byte));
    block[4] = checksum;
    Ok(())
}

fn udf_crc16(data: &[u8]) -> u16 {
    let mut crc = 0u16;
    for &byte in data {
        crc ^= u16::from(byte) << 8;
        for _ in 0..8 {
            if crc & 0x8000 != 0 {
                crc = (crc << 1) ^ 0x1021;
            } else {
                crc <<= 1;
            }
        }
    }
    crc
}

fn align_up(value: usize, align: usize) -> Option<usize> {
    debug_assert!(align.is_power_of_two());
    value
        .checked_add(align - 1)
        .map(|value| value & !(align - 1))
}

fn div_round_up(value: u64, divisor: u64) -> u64 {
    if divisor == 0 {
        return 0;
    }
    value.saturating_add(divisor - 1) / divisor
}
